flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackendBuilder.java

```diff
@@ -0,0 +1,490 @@
+		int numberOfKeyGroups,
+		KeyGroupRange keyGroupRange,
+		ExecutionConfig executionConfig,
+		LocalRecoveryConfig localRecoveryConfig,
+		RocksDBStateBackend.PriorityQueueStateType priorityQueueStateType,
+		TtlTimeProvider ttlTimeProvider,
+		MetricGroup metricGroup) {
+
+		super(kvStateRegistry, keySerializer, userCodeClassLoader,
+			numberOfKeyGroups, keyGroupRange, executionConfig, ttlTimeProvider);
+
+		this.operatorIdentifier = operatorIdentifier;
+		this.priorityQueueStateType = priorityQueueStateType;
+		this.localRecoveryConfig = localRecoveryConfig;
+		// ensure that we use the right merge operator, because other code relies on this
+		this.columnFamilyOptions = Preconditions.checkNotNull(columnFamilyOptions)
+			.setMergeOperatorName(MERGE_OPERATOR_NAME);
+		this.dbOptions = dbOptions;
+		this.instanceBasePath = instanceBasePath;
+
+		this.kvStateInformation = new LinkedHashMap<>();
+		this.keyGroupPrefixBytes =
+			RocksDBKeySerializationUtils.computeRequiredBytesInKeyGroupPrefix(numberOfKeyGroups);
+		this.writeOptions = new WriteOptions().setDisableWAL(true);
+		this.instanceRocksDBPath = new File(instanceBasePath, DB_INSTANCE_DIR_STRING);
+		this.dbPath = instanceRocksDBPath.getAbsolutePath();
+		this.rocksDBResourceGuard = new ResourceGuard();
+		this.metricGroup = metricGroup;
+		this.cancelStreamRegistry = new CloseableRegistry();
+	}
+
+	@Override
+	public RocksDBKeyedStateBackendBuilder<K> setRestoreStateHandles(Collection<KeyedStateHandle> restoreStateHandles) {
+		this.restoreStateHandles = restoreStateHandles;
+		return this;
+	}
+
+	public RocksDBKeyedStateBackendBuilder<K> setEnableIncrementalCheckpointing(boolean enableIncrementalCheckpointing) {
+		this.enableIncrementalCheckpointing = enableIncrementalCheckpointing;
+		return this;
+	}
+
+	public RocksDBKeyedStateBackendBuilder<K> setNativeMetricOptions(RocksDBNativeMetricOptions nativeMetricOptions) {
+		this.nativeMetricOptions = nativeMetricOptions;
+		return this;
+	}
+
+	public RocksDBKeyedStateBackendBuilder<K> setNumberOfTransferingThreads(int numberOfTransferingThreads) {
+		this.numberOfTransferingThreads = numberOfTransferingThreads;
+		return this;
+	}
+
+	public RocksDBKeyedStateBackendBuilder<K> setDb(RocksDB db) {
+		this.db = db;
+		return this;
+	}
+
+	private static void checkAndCreateDirectory(File directory) throws IOException {
+		if (directory.exists()) {
+			if (!directory.isDirectory()) {
+				throw new IOException("Not a directory: " + directory);
+			}
+		} else if (!directory.mkdirs()) {
+			throw new IOException(String.format("Could not create RocksDB data directory at %s.", directory));
+		}
+	}
+
+	public RocksDBKeyedStateBackend<K> build() throws IOException {
+		RocksDBKeyedStateBackend<K> backend = null;
+		RocksDBWriteBatchWrapper writeBatchWrapper = null;
+		ColumnFamilyHandle defaultColumnFamilyHandle = null;
+		RocksDBNativeMetricMonitor nativeMetricMonitor = null;
+		try {
+			prepareDirectories();
```

Comment:
The plan was to make `prepareDirectories` self-consistent. Currently the method will throw exception in three cases: 1. the base directory exists; 2. failed to create the base directory; 3. failed to delete the base directory; and adding another cleanup in `build` is kinda retry. But yes, I agree to add a "double insurance" here instead of depending on `prepareDirectories` logic.